src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccountPrimitives.fs

```diff
@@ -0,0 +1,39 @@
+﻿module Deciders.BankAccountPrimitives
+
+open System
+
+[<Struct>]
+type AccountId = AccountId of Guid
+
+module AccountId =
+    let newId () = AccountId(Guid.NewGuid())
```

Comment:
This could also just as easily be a member on the `AccountId` type itself:

```fsharp
type AccountId =
    | AccountId of Guid
    static member NewId () = AccountId (Guid.NewGuid())
```

etc.

In practice it makes little difference but (personal preference here) I like to see such functions like this (which are very closely coupled to the type itself) live on the type.

Reply:
I used FSharp.UMX per @bartelink suggestion. And it seems that I am not able to use such syntax there (probably because they're not structs?).

Reply:
Isaac definitely makes a good point if you're staying with Struct or class. I'd temper it by saying that IME for id types in domain models it winds up not staying in this mode (stuff is not really a cohesive set of things you want on the type itself but instead a set of contextually dependent mapping functions that I'd put in their own cubby holds).

With UMX, its an annotation ([ab]using the units of measure feature) that only the F# compiler knows about (the assembly has attributes on it that let the F# compiler consume it correctly, but e.g in C# it will be identical to the underlying type with no guards of any kind). So e.g. you can't do `typeof` meaningfully and GetType will say its a `System.Guid`.

To my mind thats exactly what you want for an event body type that you're going to serialize to JSON and/or if you are going to recieve them via JSON binding in ASP.NET.

There are definitely people who go the other direction and are only interested in achieving hard guarantees that no bad values can leak in. Using Struct means you've already given that up.

Lots of detail and context on these tradeoffs in the excellent series by Paul https://paul.blasuc.ci/posts/really-scu.html

You can also use `type` syntax to do these sorts of guards, but things can very longwinded for no real benefit (we had very large models and there is actually a significant perf payoff in keeping the GUIDs as strings, or there was on Full FW anyway)
https://github.com/jet/FsCodec/blob/master/tests/FsCodec.NewtonsoftJson.Tests/Fixtures.fs#L28-L47

(I personally would still select UMX for ids over struct or class SCDUs even if not using reflection/json serializers, but there's definitely people all along the spectrum)

Reply:
@bartelink, thanks for links and explanation :+1: